### ISSUE TITLE
Implementing Issue Forms and Templates

### ISSUE THREAD

**Issue by WarpWing** [FarmData2/FarmData2#229]
It would be nice to implement FarmData2 relevant Issue and PR forms that contributors can fill out to fully develop the context of their suggestion/implementation to an issue. I would recommend the following as a baseline template with specifics TBD:

Issues:
- Descriptive title
- Possible labels
- Paragraph body
- Consideration of externalities (Does it break any other components or require any tests to be run)
- (Optional) Checklist to resolve issues
- (Optional) Any existing PRs that already reference any related or directly to the issue

Pull Requests:
- Descriptive title
- Commit type
- Commit scope
- PR body
- Issues the PR closes.
- Consideration of externalities (Does it break any other components or require any tests to be run)
- (Optional) Breaking changes if present

I would like to note that with the PR Request portion, I would like to consider integration with #126 and #136. This would make PR parsing a lot easier by having contributors following a set like [this](https://github.com/FarmData2/FarmData2/issues/126#issuecomment-1884954003). It would require a lot less validation and parsing for the squashMerge script and would make it easier to generate conventional commits.

**Comment by WarpWing** [comment]
Also, I think I'll stop working on the squashMerge script as once this gets validated, it'll only take a few changes to normalize the script to accept the template.

**Comment by braughtg** [comment]
@WarpWing One question related to combining this with #126.  If templates are in place, is it possible to force compliance with the template? If not, then the squash merge script for #126 should probably only rely on what is guaranteed (title, body, etc).

**Comment by WarpWing** [comment]
> @WarpWing One question related to combining this with #126. If templates are in place, is it possible to force compliance with the template? If not, then the squash merge script for #126 should probably only rely on what is guaranteed (title, body, etc).

Yes, the issue templates allow maintainers to place form elements statically (we get to build our own forms with drop down menus and text input fields. It's all built into the form itself). I'll create an issue form to demonstrate on the squashMerge repo.

**Comment by WarpWing** [comment]
> @WarpWing One question related to combining this with #126. If templates are in place, is it possible to force compliance with the template? If not, then the squash merge script for #126 should probably only rely on what is guaranteed (title, body, etc).

@braughtg  I've created a example issue template from the documentation. You can check it out by creating a new issue here: https://github.com/FarmData2/squashMergeTest/issues. This is customizable, just an example.

**Comment by WarpWing** [comment]
Ahhhh, I think I'm mistaken. I think that these forms are for issues only, not for PRs. However, we can still use standard PR templates (outside of the beta static issue forms) and I can still normalize the PR script via tracking headers or the titles of each subbody. We could technically kinda force compliance since the PR template will appear by default but we can't strictly do it since authors could delete the template.

**Comment by braughtg** [comment]
@WarpWing I had a look at the bug report form example and I really like the idea.

One concern I have:
- The forms are listed as "in beta" and may change.  Do you have a sense of how stable they are? While I think the forms will be great, I hesitate to get in too early and create things that will break frequently and thus require lots of attention to keep them working while they stabilize.  This is purely a time management concern.

Aside from that concern, I think we will eventually have these (now or when they are more stable) so we can still consider what forms we will want and what should be on those forms.  
- I think the place to start is to do some research into the types of templates projects use or any articles we can find on best practices or example templates (from GitHub or any other tracker that we could adapt).
- We can collect that information in this issue.

As a start:
- We might try to find sample templates to adapt for:
  - Bug report (separate for in application vs dev environment?)
    - Here is the article that students read in COMP190 that talks about [How to write effective bug reports](https://capgemini.github.io/testing/effective-bug-reports/). It could be a good starting point for a bug report template. 
  - User facing features (possibly using a user story model with acceptance tests?)
    - Is there a difference between a new feature and modification of an existing feature?
  - Developer facing features
  - Tests to be written for existing code (unit, component, e2e)
  - Documentation to be written
  - Performance improvements
  - Code improvements (refactoring, consistency, idioms)
  - Others?

**Comment by WarpWing** [comment]
> @WarpWing I had a look at the bug report form example and I really like the idea.
> 
> One concern I have:
> 
>     * The forms are listed as "in beta" and may change.  Do you have a sense of how stable they are? While I think the forms will be great, I hesitate to get in too early and create things that will break frequently and thus require lots of attention to keep them working while they stabilize.  This is purely a time management concern.
> 
> 
> Aside from that concern, I think we will eventually have these (now or when they are more stable) so we can still consider what forms we will want and what should be on those forms.
> 
>     * I think the place to start is to do some research into the types of templates projects use or any articles we can find on best practices or example templates (from GitHub or any other tracker that we could adapt).
> 
>     * We can collect that information in this issue.
> 
> 
> As a start:
> 
>     * We might try to find sample templates to adapt for:
>       
>       * Bug report (separate for in application vs dev environment?)
>         
>         * Here is the article that students read in COMP190 that talks about [How to write effective bug reports](https://capgemini.github.io/testing/effective-bug-reports/). It could be a good starting point for a bug report template.
>       * User facing features (possibly using a user story model with acceptance tests?)
>         
>         * Is there a difference between a new feature and modification of an existing feature?
>       * Developer facing features
>       * Tests to be written for existing code (unit, component, e2e)
>       * Documentation to be written
>       * Performance improvements
>       * Code improvements (refactoring, consistency, idioms)
>       * Others?

@braughtg Currently, the Issue Forms are indeed listed as beta but have been for approximately 2 years now according to this[ article](https://github.blog/changelog/2021-06-23-issues-forms-beta-for-public-repositories/) date so it's pretty stable. I've seen other repositories use it and from the feedback form, it seems they are focused on additions rather than core fixes so it looks pretty stable but I'll let you make the final decision.


In the meantime, I'll start creating a template prototypes and I'll try to fully develop them and look into your suggestions. I definitely want a comprehensive form but I also don't want to exhaust the developer too hard. 

This does give me an idea to perhaps create a Github Action to automatically spawns issues from merged PRs that contain subsequent work (eg. someone Pull Requests a new component that gets accepted but under a certain section, if indicated, the component requires documentation and refactoring and the bot could spawn those issues with labels for others to work on).

I also want to expand upon this:
>         * Is there a difference between a new feature and modification of an existing feature?
>       * Developer facing features
>       * Tests to be written for existing code (unit, component, e2e)
>       * Documentation to be written
>       * Performance improvements
>       * Code improvements (refactoring, consistency, idioms)
>       * Others?

As I would like to also create conditionals as in if the issue indicates that it's modifying an existing component to have a bot or some system chime in that "You should also make sure to check against the E2E or component tests for X and update the documentation for X". Personally, I've been admittedly been guilty of not considering other factors or components when updating or writing code for a component or script.
